server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java

```diff
@@ -0,0 +1,156 @@
+/*
+* Licensed to Metamarkets Group Inc. (Metamarkets) under one
+* or more contributor license agreements. See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership. Metamarkets licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License. You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.segment.realtime.plumber;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.metamx.common.Pair;
+import com.metamx.common.logger.Logger;
+import io.druid.client.SegmentLoadInfo;
+import io.druid.client.coordinator.CoordinatorClient;
+import io.druid.concurrent.Execs;
+import io.druid.query.SegmentDescriptor;
+import io.druid.server.coordination.DruidServerMetadata;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class CoordinatorBasedSegmentHandoffNotifier implements SegmentHandoffNotifier
+{
+  private static final Logger log = new Logger(CoordinatorBasedSegmentHandoffNotifier.class);
+
+  private final Map<SegmentDescriptor, Pair<Executor, Runnable>> handOffCallbacks = Maps.newConcurrentMap();
+  private final CoordinatorClient coordinatorClient;
+  private volatile ScheduledExecutorService scheduledExecutor;
+  private final long pollDurationMillis;
+  private final String dataSource;
+
+  public CoordinatorBasedSegmentHandoffNotifier(
+      String dataSource,
+      CoordinatorClient coordinatorClient,
+      CoordinatorBasedSegmentHandoffNotifierConfig config
+  )
+  {
+    this.dataSource = dataSource;
+    this.coordinatorClient = coordinatorClient;
+    this.pollDurationMillis = config.getPollDuration().getMillis();
+  }
+
+  @Override
+  public void registerSegmentHandoffCallback(
+      SegmentDescriptor descriptor, Executor exec, Runnable handOffRunnable
+  )
+  {
+    log.info("Adding SegmentHandoffCallback for dataSource[%s] Segment[%s]", dataSource, descriptor);
+    handOffCallbacks.put(descriptor, new Pair<>(exec, handOffRunnable));
+  }
+
+  @Override
+  public void start()
+  {
+    scheduledExecutor = Execs.scheduledSingleThreaded("coordinator_handoff_scheduled_%d");
+    scheduledExecutor.scheduleAtFixedRate(
+        new Runnable()
```

Comment:
Always make sure to wrap the entire `Runnable` that you are scheduling in a try/catch if you want it to actually continue running.  If an exception *ever* makes it out of the `Runnable`, the schedule will halt.  You have a try/catch over some of the business logic, but what happens if you get an `NPE` on 80 or for some reason `itr.next()` throws an exception or something.